### PR TITLE
#152 (Partition TOSA at conv2d) and #162 (Element-wise attribute for TOSA ops)

### PR DESCRIPTION
**Pull request by pcf000**
Look for tosa::Conv2D ops followed and/or preceded by element-wise ops, and pull them out into separate functions with a "kernel" attribute for later fusion.  Requires the invention of an element-wise attribute.

Also update some integration  tests for the external/llvm-project update.

### CODE REVIEW

File: external/llvm-project/mlir/test/Integration/GPU/ROCM/gpu-to-hsaco.mlir [hunk 1]
```diff
@@ -11,23 +11,23 @@
 func @other_func(%arg0 : f32, %arg1 : memref<?xf32>) {
   %c0 = constant 0 : index
   %c1 = constant 1 : index
-  %block_dim = dim %arg1, %c0 : memref<?xf32>
+  %block_dim = memref.dim %arg1, %c0 : memref<?xf32>
```
Comment (pcf000) [inline external/llvm-project/mlir/test/Integration/GPU/ROCM/gpu-to-hsaco.mlir:14]: Some dialect movements.  These tests don't run unless -DMLIR_INCLUDE_INTEGRATION_TESTS=ON.  I found them by accident when I put my resnet50 whole-model test in with the integration tests.

File: external/llvm-project/mlir/test/Dialect/Tosa/tosa-partition-run.mlir [hunk 2]
```diff
@@ -0,0 +1,46 @@
+// RUN: mlir-opt %s --tosa-to-linalg-on-tensors --tosa-to-standard --linalg-detensorize \
```
Comment (pcf000) [inline external/llvm-project/mlir/test/Dialect/Tosa/tosa-partition-run.mlir:1]: This is a simple arbitrary test of a convolution followed by a RelU, generated from tensorflow with known inputs so we can test both that it got the right answer and that --tosa-partition-pipeline didn't affect the answer.

File: external/llvm-project/mlir/tools/mlir-opt/mlir-opt.cpp [hunk 3]
```diff
@@ -106,6 +106,7 @@ void registerTestPreparationPassWithAllowedMemrefResults();
 void registerTestRecursiveTypesPass();
 void registerTestSCFUtilsPass();
 void registerTestVectorConversions();
+void registerTosaPartitionPass();
```
Comment (pcf000) [inline external/llvm-project/mlir/tools/mlir-opt/mlir-opt.cpp]: It doesn't have to be in with the test options, but what I was copying from did that so I did too.

File: CMakeLists.txt [hunk 4]
```diff
@@ -40,6 +40,10 @@ set(LLVM_ENABLE_TERMINFO OFF CACHE BOOL "")
 # LLVM settings that have an effect on the MLIR dialect
 set(LLVM_TARGETS_TO_BUILD "X86;AMDGPU" CACHE STRING "")
 
+if (MLIR_INCLUDE_TESTS)
```
Comment (pcf000) [inline CMakeLists.txt]: I think we end up having to use -DMLIR_INCLUDE_TESTS to get --tosa-partition-pipeline in our mlir-opt, though not in the external mlir-opt.

**Comment by pcf000** [comment]
We can talk about this on Tuesday, if you'd like.

**Comment by whchung** [comment]
@pcf000 Need `clang-format` on this PR:

http://10.236.106.97:8080/blue/organizations/jenkins/mlir/detail/PR-412/6/pipeline

**Comment by pcf000** [comment]
On the one hand, yes, I haven't run clang-format.
On the other hand, I didn't modify mlir/include/mlir/InitMIOpenDialects.h that it complains about.  In fact, it doesn't exist on my branch;  Krzysztof added it last night.

**Comment by whchung** [comment]
@pcf000 I think that `clang-format` check is applied by merging your branch with `miopen-dialect`. And for some reason @krzysz00 's PR went in without bumping into the issue.

One thing I'd suggest is to do one of the followings in your local branch:

1) `git clang-format HEAD~X`, X be a number large enough than your commits in this branch.
2) `git merge miopen-dialect`, followed by `git clang-format HEAD^`.
3) `git rebase miopen-dialect`, followed by `git clang-format HEAD~X`, X be a number large enough than your commits in this branch.

**Comment by whchung** [comment]
@pcf000 ah... I see you have just made a commit which did the `clang-format` thing. please disregard my earlier comment then.

**Comment by pcf000** [comment]
We were thinking the same thing at the same time.

**Comment by pcf000** [comment]
@sjw36, most of what we talked about is present here -- the partition passes have moved out of the test namespace and thus everything is under external, I've run clang-format, and I'm comparing attributes in isFunctionallyEquivalentTo().

Now I'm looking into adapting mergeIdenticalBlocks() into something that collapses identical outlined kernels while gathering them rather than after they're outlined;  it happens to already use a hash function, too.  When that's done, or if it becomes too complex, I'll call this PR ready to approve.

File: external/llvm-project/mlir/include/mlir/Dialect/Tosa/IR/TosaOpBase.td [hunk 5]
```diff
@@ -192,4 +192,9 @@ class Tosa_Op<string mnemonic, list<OpTrait> traits = []> :
     Op<Tosa_Dialect, mnemonic, !listconcat(traits, [TosaOpInterface])> {
 }
 
+//===----------------------------------------------------------------------===//
+// TOSA element-wise ops.  The Elementwise trait requires identical shapes.
```
Comment (sjw36) [inline external/llvm-project/mlir/include/mlir/Dialect/Tosa/IR/TosaOpBase.td:196]: We will probably want to support broadcast such that:

> tosa.add(%arg0 : tensor<128x32x32x64xf32>, %arg2 : tensor<1x1x1x64xf32>)

would broadcast all dims of size 1. This is similar to bias-add.

Not necessary for this PR.

File: external/llvm-project/mlir/include/mlir/Dialect/Tosa/IR/TosaOps.td [hunk 6]
```diff
@@ -7,7 +7,7 @@
 //===----------------------------------------------------------------------===//
 //
 // This file defines the operation set for the TOSA dialect as defined in
-// the TOSA specfication (https://developer.mlplatform.org/w/tosa/). 
+// the TOSA specfication (https://developer.mlplatform.org/w/tosa/).
```
Comment (sjw36) [inline external/llvm-project/mlir/include/mlir/Dialect/Tosa/IR/TosaOps.td]: still has space diffs.

File: mlir/include/mlir/InitMIOpenDialects.h [hunk 7]
```diff
@@ -1,5 +1,5 @@
 //===- InitMIOpenDialects.h - MLIR MIOpen Dialects Registration -----------*-
-//C++ -*-===//
+// C++ -*-===//
```
Comment (sjw36) [inline mlir/include/mlir/InitMIOpenDialects.h]: revert this.

**Review by sjw36 — APPROVED** [review]
Looks good.

File: external/llvm-project/mlir/include/mlir/Dialect/Tosa/IR/TosaOps.td (hunk 6 above)
Comment (pcf000) [reply external/llvm-project/mlir/include/mlir/Dialect/Tosa/IR/TosaOps.td]: Hmm, latest version of that file doesn't have the space, but does correct the spelling of "specification."

File: mlir/include/mlir/InitMIOpenDialects.h (hunk 7 above)
Comment (pcf000) [reply mlir/include/mlir/InitMIOpenDialects.h]: And this change is no longer present, having been done in a separate pull request.

**Pull request merged**